add(scan): Adds gRPC reflection and documents how to use the zebra-scan gRPC server (#8288)
ZcashFoundation/zebra: book/src/user/shielded-scan-grpc-server.md

# Zebra Shielded Scanning gRPC Server

## Get Started

### Setup

> After setting up Zebra Shielded Scanning, add a `listen_addr` field to the shielded-scan configuration:

Reviewer: ```suggestion
After setting up [Zebra Shielded Scanning](shielded-scan.md), add a `listen_addr` field to the shielded-scan configuration:
```

Author: Thank you, applied in https://github.com/ZcashFoundation/zebra/pull/8288/commits/6a1057ae5418a1afebdf89d895f5f3516cd6466f.

```toml
[shielded_scan]
listen_addr = "127.0.0.1:8231"
```

Then, run `zebrad` to start the scan gRPC server.

Making requests to the server will also require a gRPC client, the examples here use `grpcurl`, though any gRPC client should work.

[See installation instructions for `grpcurl` here](https://github.com/fullstorydev/grpcurl?tab=readme-ov-file#installation).

The types can be accessed through the `zebra-grpc` crate's root `scanner` module for clients in a Rust environment, and the [`scanner.proto` file here](https://github.com/ZcashFoundation/zebra/blob/main/zebra-grpc/proto/scanner.proto) can be used to build types in other environments.

### Usage

To check that the gRPC server is running, try calling `scanner.Scanner/GetInfo`, for example with `grpcurl`:

```bash

> $ grpcurl -plaintext '127.0.0.1:8231' scanner.Scanner/GetInfo

Reviewer: ```suggestion
grpcurl -plaintext '127.0.0.1:8231' scanner.Scanner/GetInfo
```

We have some bash commands starting with `$` while others don't. For consistency lets do all the same, i prefer without it so command can be copy-pasted.

Author: Applied in https://github.com/ZcashFoundation/zebra/pull/8288/commits/6a1057ae5418a1afebdf89d895f5f3516cd6466f.

```

The response should look like:

```
{
  "minSaplingBirthdayHeight": 419200
}
```

An example request to the `Scan` method with `grpcurl` would look like: 

```bash

> grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/Scan

Reviewer: ```
alfredo@spaceship:~/zebra$ ./grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/Scan
Error invoking method "scanner.Scanner/Scan": service "scanner.Scanner" does not include a method named "Scan"
alfredo@spaceship:~/zebra$ 
```

Author: This worked for me just now after rebasing on main:

```console

grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/Scan
ERROR:
  Code: Unknown
  Message: scan service returned error: channel closed
```

Reviewer: I think the problem might be if the state is below the sapling activation or similar. When i use the command i get a hang (this is more than 10 or 30 seconds, it seems to be something else).

Reviewer: Ok, i have more details in the hangs. If we are below sapling activation the grpc methods that require any state will hang. I can see this messages in zebrad logs:

```
zebra_scan::service::scan_task::scan: scanner is waiting for Sapling activation. Current tip: 106951, Sapling activation: 419200
```

So we need to return an error from the grpc call if we are below sapling activation height. We can do this in another PR.

Reviewer: Should we add an error about the invalid key used?

```bash
alfredo@spaceship:~/zebra$ ./grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/Scan
ERROR:
  Code: Unknown
  Message: scan service returned error: channel closed
alfredo@spaceship:~/zebra$
```

Why this is not working? the error doesn't tell us too much.
```
alfredo@spaceship:~/zebra$ ./grpcurl -plaintext -d '{ "keys": { "key": ["zxviews1q0duytgcqqqqpqre26wkl45gvwwwd706xw608hucmvfalr759ejwf7qshjf5r9aa7323zulvz6plhttp5mltqcgs9t039cx2d09mgq05ts63n8u35hyv6h9nc9ctqqtue2u7cer2mqegunuulq2luhq3ywjcz35yyljewa4mgkgjzyfwh6fr6jd0dzd44ghk0nxdv2hnv4j5nxfwv24rwdmgllhe0p8568sgqt9ckt02v2kxf5ahtql6s0ltjpkckw8gtymxtxuu9gcr0swvz"] } }' '127.0.0.1:8231' scanner.Scanner/Scan
ERROR:
  Code: Unknown
  Message: scan service returned error: channel closed
alfredo@spaceship:~/zebra$ 
```

Reviewer 2: Yea, I also think the hang is not a blocker for this PR. It would be good to get it merged before the release.

Reviewer: I opened a separated ticket to work on those (https://github.com/ZcashFoundation/zebra/issues/8316). Feel free to close all conversations and merge.

```

This will start scanning for transactions in Zebra's state and in new blocks as they're validated.

> Or, to use the scanner gRPC server without streaming, try calling `RegisterKeys` with your Sapling extended full viewing key, waiting for the scanner to cache some results, then calling `GetResults`.

Reviewer: I tried the next command and it seems to hang there without returning. Can you double check using this pull request code ?

```suggestion
Or, to use the scanner gRPC server without streaming, try calling `RegisterKeys` with your Sapling extended full viewing key, waiting for the scanner to cache some results, then calling `GetResults`:

```bash
grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/RegisterKeys
```

Author: It should hang there briefly until the `RegisterKeys` messages are processes, it's faster on main since the `CHECK_INTERVAL` was reduced to 10 seconds, I just tested it after rebasing this PR and it returned relatively quickly:

```console

grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/RegisterKeys
{}
```

Added the example in https://github.com/ZcashFoundation/zebra/pull/8288/commits/6a1057ae5418a1afebdf89d895f5f3516cd6466f.

## gRPC Reflection

To see all of the provided methods with `grpcurl`, try:

```bash
$ grpcurl -plaintext '127.0.0.1:8231' list scanner.Scanner
```

This will list the paths to each method in the `Scanner` service:
```
scanner.Scanner.ClearResults
scanner.Scanner.DeleteKeys
scanner.Scanner.GetInfo
scanner.Scanner.GetResults
scanner.Scanner.RegisterKeys
```

To see the the request and response types for a method, for example the `GetResults` method, try:

```bash

> $ grpcurl -plaintext '127.0.0.1:8231' describe scanner.Scanner.GetResults \

Reviewer: ```suggestion
grpcurl -plaintext '127.0.0.1:8231' describe scanner.Scanner.GetResults \
```

Author: Applied in https://github.com/ZcashFoundation/zebra/pull/8288/commits/6a1057ae5418a1afebdf89d895f5f3516cd6466f.

# Zebra Shielded Scanning gRPC Server

## Get Started

### Setup

> After setting up [Zebra Shielded Scanning](shielded-scan.md), add a `listen_addr` field to the shielded-scan configuration:

Reviewer 2: Will the `shielded-scan.md` link point to the right resource?

If not, let's replace it with:
```suggestion
After setting up [Zebra Shielded Scanning](https://zebra.zfnd.org/user/shielded-scan.html), add a `listen_addr` field to the shielded-scan configuration:
```

Reviewer: Thanks for the fix, the previous will work in the github but not in the book,

The types can be accessed through the `zebra-grpc` crate's root `scanner` module for clients in a Rust environment, and the [`scanner.proto` file here](https://github.com/ZcashFoundation/zebra/blob/main/zebra-grpc/proto/scanner.proto) can be used to build types in other environments.

### Usage

To check that the gRPC server is running, try calling `scanner.Scanner/GetInfo`, for example with `grpcurl`:

```bash
grpcurl -plaintext '127.0.0.1:8231' scanner.Scanner/GetInfo
```

The response should look like:

```
{
  "minSaplingBirthdayHeight": 419200
}
```

An example request to the `Scan` method with `grpcurl` would look like: 

```bash
grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/Scan
```

This will start scanning for transactions in Zebra's state and in new blocks as they're validated.

Or, to use the scanner gRPC server without streaming, try calling `RegisterKeys` with your Sapling extended full viewing key, waiting for the scanner to cache some results, then calling `GetResults`:

```bash

> grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/RegisterKeys

Reviewer: It seems this just works even if the sapling key is invalid?

```
alfredo@spaceship:~/zebra$ ./grpcurl -plaintext -d '{ "keys": { "key": ["sapling_extended_full_viewing_key"] } }' '127.0.0.1:8231' scanner.Scanner/RegisterKeys
{}
alfredo@spaceship:~/zebra$ 
```

We get the same responseif the key is valid.
```
alfredo@spaceship:~/zebra$ ./grpcurl -plaintext -d '{ "keys": { "key": ["zxviews1q0duytgcqqqqpqre26wkl45gvwwwd706xw608hucmvfalr759ejwf7qshjf5r9aa7323zulvz6plhttp5mltqcgs9t039cx2d09mgq05ts63n8u35hyv6h9nc9ctqqtue2u7cer2mqegunuulq2luhq3ywjcz35yyljewa4mgkgjzyfwh6fr6jd0dzd44ghk0nxdv2hnv4j5nxfwv24rwdmgllhe0p8568sgqt9ckt02v2kxf5ahtql6s0ltjpkckw8gtymxtxuu9gcr0swvz"] } }' '127.0.0.1:8231' scanner.Scanner/RegisterKeys
{}
alfredo@spaceship:~/zebra$ 
```

Author: It returns a list of keys that were registered, maybe it should return an error if no keys were added?

Reviewer: Yea ... lets work on it in the context of https://github.com/ZcashFoundation/zebra/issues/8316
